src/client/rpcIdentities.ts

```diff
@@ -26,17 +26,16 @@ const createIdentitiesRPC = ({
 }) => {
   return {
     identitiesAuthenticate: async (
```

Comment:
I forgot how this call is supposed to work. Are we expecting a back and forth protocol between the CLI user and the agent? Or are you saying that the user provides the token upon calling authenticate?

I thought that the provider has to start the operation, and then give you a URL that you have to go to, and then the user has to copy that URL back into the CLI. That's why the `authenticate` is an async generator as it represents this particular flow of interaction. Therefore you would need a duplex stream right?

Reply:
I believe the authenticate call needs to be a duplex stream. For `TestProvider` it doesn't matter. But for `GitHubProvider` it does matter and all real providers will need to be an "interaction". This will need to convert the duplex stream to using the `prompts` library to prompt the user for more input.

Reply:
I need to refresh my memory of how the authentication process actually works for our actual providers (i.e. not the `TestProvider`). Going to write a test.ts and re-augment my GitHub to map out the process that will be required for the stream.

The current test in `rpcIdentities.test.ts` assumes the easiest case of the `TestProvider`, where no back and forth is required:

```ts
  test('should authenticate an identity', async () => {
    const identitiesAuthenticate =
      grpcUtils.promisifyUnaryCall<identitiesPB.Token>(
        client,
        client.identitiesAuthenticate,
      );

    const providerMessage = new identitiesPB.Provider();
    providerMessage.setProviderId(testToken.providerId);
    providerMessage.setIdentityId(testToken.identityId);

    const codeMessage = await identitiesAuthenticate(
      providerMessage,
      callCredentials,
    );
    expect(codeMessage.getToken()).toContain('randomtestcode');
    expect(
      await polykeyAgent.identitiesManager.getToken(
        testToken.providerId,
        testToken.identityId,
      ),
    ).toEqual(testToken.tokenData);
    await polykeyAgent.identitiesManager.delToken(
      testToken.providerId,
      testToken.identityId,
    );
  });
```

and `TestProvider.authenticate` is the following:

```ts
  public async *authenticate(): AsyncGenerator<string | undefined, IdentityId> {
    const code = 'randomtestcode';
    yield code;
    // Always gives back the abc123 token
    const tokenData = { accessToken: 'abc123' };
    const identityId = await this.getIdentityId(tokenData);
    await this.putToken(identityId, tokenData);
    return identityId;
  }
```

Reply:
So our `authenticate` function has the following flow:
1. We perform a request to GitHub (or some provider)
2. This produces a usercode, which is yielded to the user.
3. The user manually inputs this usercode into the opened GitHub URL
4. `authenticate` performs the remainder of the authentication. It polls on this authentication request, waiting for the usercode to be manually entered
5. After it's been entered, it finally returns the token data/identity ID

Reply:
I actually don't think there's an issue with this? I've recorded myself performing a GitHub authentication with the current `authenticate` CLI command. Is this not what you expect @CMCDragonkai?

https://user-images.githubusercontent.com/49772421/146286560-f7d08772-dfd9-4832-b28f-665f66fcdf71.mp4

Reply:
If this is a unary call how do you get the user code to the user?

Reply:
I mean the URL?

Reply:
Remember the agent has to tell the user to visit a URL. If the client is on a different host. They would need a URL to access. The CLI client can auto open a browser. But if it fails to do so, the url should be printed out to STDOUT and then prompt for the code?

Reply:
I just watched the video. There are several things we need to change there:

1. The opening the browser to the authentication URL should be done on the `CommandAuthenticate.ts` which means it can execute on the client side even if the PK agent exists on a different host. The video shows this being executed by the server side which is not a good idea.
2. The automatic opening of the browser may fail because it relies on the platform specifics which may not work. This means the target URL should be displayed on the CLI side, and if the browser fails to run, it will just display the URL for the user to visit directly.
3. The user code is provided to the user to enter into the website being opened up. The message should be "Enter this Code: ". Sort of a key-value structure for the STDOUT.
4. The entire CLI call should probably only terminate when the entire authentication process is done. This ensures one can cancel the entire process if need be.

In order to solve 1./2. the `authenticate` may need to change a bit. Right now it directly calls `browser()` which means this executes on the agent side. This would be something that the CLI command would instead execute.

On the agent side, it can continue pinging the `https://github.com/login/oauth/access_token` until it acquires the token. And if it succeeds it puts that token against the identity ID.

So here's what I propose:

1. The `authenticate` process yields 2 things: the user code and the URL you're supposed to go to on your browser.
2. The GRPC handler provides 2 things in its stream. It's important to be a stream, because the last message in the stream SHOULD be the identity id.
3. The stream should a server stream. So you know when the authentication process has successfully completed. This allows us to communicate the agent to cancel this operation if we need to.
4. The `CommandAuthenticate.ts` takes the URL and the user code (which may be optional). For example: `public async *authenticate(timeout: number = 120000): AsyncGenerator<string | undefined, IdentityId>;` is the signature for `authenticate`. What this means that the `string | undefined` was originally for the user code. Some providers may not do this at all. Instead we need change this to `AsyncGenerator<{url: string; code?: string}, IdentityId>`. Note that the `IdentityId` is what is returned at the end.
5. The `CommandAuthenticate` uses the `browser` function to try and launch the `url`. Whether this fails or not, it will show the url on the STDOUT as well. If it does fail, a STDERR message should be provided to tell the user to try and visit the URL. The code is optional because some providers don't do this. But if it does exist, it should be some like `Enter this Code: ...`.
6. At this point, the `CommandAuthenticate` will just `await` on the stream to be finished and acquire the `IdentityId`. While this is happening, in the agent, it's polling the authentication URL to see if the authentication process has been completed. If it completes, it will then return the identity id, and this will close the stream. The `CommandAuthenticate` can then output the authenticated ID in its STDOUT.
7. This means that the CLI execution only finishes when authentication process completes.

There are some exceptions:

1. If the timeout times out, the agent actually throws the exception, this exception is thrown to the stream which should be done in the handler, it is propagated to the client. The client then outputs the error and exits the process.
2. If the authentication fails, as in the user didn't provide the right user code or if the authentication was rejected, then an exception will be propagated to the client as well.

Reply:
So try this instead:

```
AsyncGenerator<{ url: string; [key: string]: string; }, IdentityId>
```

The only issue with this is the fact that because it's any, you have to convert to a string. So let's just keep it as a string.

Then on the GRPC protobuf, you would need an a message with arbitrary number of data. This could be a map. There's always the JSON encode. You could try using the `Struct` or directly with `map`.

Reply:
On the CLI side, you can use the output formatter for an arbitrary dictionary to display.

Reply:
Given that we're using `[key: string]: string;` in `AsyncGenerator<{ url: string; [key: string]: string; }, IdentityId>`, we can enforce this at the proto level too.

i.e. instead of using the `Struct` type (that supports `Value`s of any type), we can just use a `map<string, string>` for the extra data.

Reply:
Remaining TODOs for this:
- [ ] - add required protobuf messages
- [ ] - adjust corresponding GRPC calls/signatures
- `GithubProvider`:
   * [ ] - adjust async generator return types
   * [ ] - remove `browser` call
- `rpcIdentities.identitiesAuthenticate`
   * [ ] - change to using the stream
   * [ ] - add support for arbitrary data with `[key: string]: string` 
   * [ ] - adjust flow to utilise the generator and corresponding calls to stream
   * [ ] - remove callbacks
   * [ ] - add exception handling (and propagate to stream closures)
- `CommandAuthenticate`:
   * [ ] - add `browser` call here
   * [ ] - adjust output to display URL, user code (where necessary) and appropriate messaging
   * [ ] - change to using the stream

Reply:
It's working for now, so this can be resolved.